src/lib.rs

```diff
@@ -852,22 +845,23 @@ impl<'a, L: 'static + Ledger> KeystoreState<'a, L> {
                 // This maintains the invariant that everything in `pending_transactions` must
                 // correspond to an on-hold record, because everything which corresponds to a record
                 // whose hold just expired will be removed from the set now.
-                for txn in self.txn_state.clear_expired_transactions() {
+                for txn in self
+                    .txn_state
+                    .clear_expired_transactions(&mut model.transactions)
+                {
                     summary
                         .updated_txns
-                        .push((txn.uid(), TransactionStatus::Rejected));
+                        .push((txn.uid().clone(), TransactionStatus::Rejected));
                     model.backend.finalize(txn, None).await;
                 }
             }
-
             LedgerEvent::Memos {
                 outputs,
                 transaction,
             } => {
                 let completed = self
                     .txn_state
-                    .transactions
-                    .received_memos(outputs.iter().map(|info| info.2));
+                    .received_memos(outputs.iter().map(|info| info.2), &mut model.transactions);
```

Comment:
Why does this go through `txn_state` at all now? The transaction _builder_ should operate only on the ledger state, not pending transactions, and indeed if I'm looking at the right thing, `TransactionState::received_memos` doesn't use `self`. It seems like this should be replaced with a direct invocation of the `Transactions` model, like you did above with `add_pending_uids`. (Sort of inlining some logic that will eventually make up the `TransactionsController`)